tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java

```diff
@@ -780,6 +759,147 @@ private ApplicationRpcServer setupRPCService(String hostname) {
+      AMRMClient.ContainerRequest containerAsk = setupContainerRequestForRM(request);
+      String jobName = request.getJobName();
+      if (!jobTypeToContainerRequestsMap.containsKey(jobName)) {
+        jobTypeToContainerRequestsMap.put(jobName, new ArrayList<>());
+        jobTypeToContainerResources.put(jobName, getContainerResources(jobName));
+      }
+      jobTypeToContainerRequestsMap.get(request.getJobName()).add(containerAsk);
+      for (int i = 0; i < request.getNumInstances(); i++) {
+        amRMClient.addContainerRequest(containerAsk);
+        session.addExpectedTask();
+      }
+    }
+
+    private AMRMClient.ContainerRequest setupContainerRequestForRM(TensorFlowContainerRequest request) {
+      Priority priority = Priority.newInstance(request.getPriority());
+      Resource capability = Resource.newInstance((int) request.getMemory(), request.getVCores());
+      Utils.setCapabilityGPU(capability, request.getGPU());
+      AMRMClient.ContainerRequest containerRequest = new AMRMClient.ContainerRequest(capability, null, null, priority, true, request.getNodeLabelsExpression());
+      LOG.info("Requested container ask: " + containerRequest.toString());
+      return containerRequest;
+    }
+
+    private Map<String, LocalResource> getContainerResources(String jobName) {
+      Map<String, LocalResource> containerResources = new ConcurrentHashMap<>(localResources);
+      String[] resources = tonyConf.getStrings(TonyConfigurationKeys.getResourcesKey(jobName));
+      Utils.addResources(resources, containerResources, resourceFs);
+
+      // All resources available to all containers
+      resources = tonyConf.getStrings(TonyConfigurationKeys.getContainerResourcesKey());
+      Utils.addResources(resources, containerResources, resourceFs);
+      return containerResources;
+    }
+
+    private synchronized void registerDependencyCompleted(String jobName) {
+      taskDependencyMap.forEach((k, v) -> {
+        if (v.containsKey(jobName)) {
+          int numContainersLeft = v.get(jobName);
+          numContainersLeft--;
+
+          if (numContainersLeft == 0) {
+            v.remove(jobName);
+          } else {
+            v.put(jobName, numContainersLeft);
+          }
+        }
+      });
+
+      Iterator<TensorFlowContainerRequest> waitingRequestItr = taskDependencyMap.keySet().iterator();
+      while (waitingRequestItr.hasNext()) {
+        TensorFlowContainerRequest waitingRequest = waitingRequestItr.next();
+        if (checkDependencySatisfied((waitingRequest))) {
+          waitingRequestItr.remove();
+          scheduleJob(waitingRequest);
+        }
+      }
+    }
+
+    private boolean isDAG(final List<TensorFlowContainerRequest> containersRequests) {
+      Map<String, TensorFlowContainerRequest> containersRequestsMap = new HashMap<>();
+      containersRequests.forEach(x -> containersRequestsMap.put(x.getJobName(), x));
+
+      for (TensorFlowContainerRequest containerRequest : containersRequests) {
```

Comment:
Seems this might be optimizable by creating the `visited` set here, passing it through to isSubgraphDAG and checking in this for loop if we've seen it before